rust/datafusion/src/physical_plan/hash_join.rs

```diff
@@ -0,0 +1,467 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the join plan for executing partitions in parallel and then joining the results
+//! into a set of partitions.
+
+use std::sync::Arc;
+use std::{
+    any::Any,
+    collections::{HashMap, HashSet},
+};
+
+use async_trait::async_trait;
+use futures::{Stream, StreamExt, TryStreamExt};
+
+use arrow::array::{make_array, Array, MutableArrayData};
+use arrow::datatypes::{Schema, SchemaRef};
+use arrow::error::Result as ArrowResult;
+use arrow::record_batch::RecordBatch;
+
+use super::hash_utils::{build_join_schema, check_join_is_valid, JoinHow};
+use super::{expressions::col, hash_aggregate::create_key};
+use crate::error::{DataFusionError, Result};
+
+use super::{
+    group_scalar::GroupByScalar, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+
+// An index of (batch, row) uniquely identifying a row in a part.
+type Index = (usize, usize);
+// None represents a null (e.g. in case of a left join, some right indices are null)
+type JoinIndex = (usize, usize);
+// A mapping "on" value -> list of row indexes with this key's value
+// E.g. [1, 2] -> [(0, 3), (1, 6), (0, 8)] indicates that (column1, column2) = [1, 2] is true
+// for rows 3 and 8 from batch 0 and row 6 from batch 1.
+type JoinHashMap = HashMap<Vec<GroupByScalar>, Vec<Index>>;
+type JoinLeftData = (JoinHashMap, Vec<RecordBatch>);
+
+/// join execution plan executes partitions in parallel and combines them into a set of
+/// partitions.
+#[derive(Debug)]
+pub struct HashJoinExec {
+    /// left side
+    left: Arc<dyn ExecutionPlan>,
+    /// right side
+    right: Arc<dyn ExecutionPlan>,
+    /// Set of common columns used to join on
+    on: HashSet<String>,
```

Comment:
This is a great start but will limit us to cases where the column names are the same between two tables. As a follow-on we can expand this to support different names e..g `customer.id = orders.customer_id`.

Reply:
After thinking about this some more, I think it might be better to address this in the current PR. I would suggest either:

```rust
left_keys: Vec<String>
right_keys: Vec<String>,
```

or the more general case:

```rust
left_keys: Vec<Arc<dyn PhysicalExpr>>
right_keys: Vec<Arc<dyn PhysicalExpr>>,
```

It should be possible to perform an equi-join using any deterministic expression, for example:

```
SELECT a.*, b.* FROM a JOIN b on UPPER(a.name) = UPPER(CONCAT(b.first, ' ', b.last))
```

Reply:
Agreed. I will have this fixed: I do not see any technical challenge here (apart from the schema changes).

Reply:
Another option here is to have an optimizer rule that whenever a non-column expression (e.g. `a.a + 1 = a.b + 2`) is passed to the `on`, it converts it to a projection on both sides. In terms of compute, AFAIK it is the same, as we always need to evaluate the array before computing the hash of each of its rows, and evaluating it on the `Join` or on a preceding projection seems to be equivalent.

The rational for using `HashSet<String>` was that this is way easier to represent (i.e. serialize and equate) on a partition key, which means when we introduce hashJoins on partitioned keys, it may be easier to tackle.

Reply:
Yes, this is an implicit projection that happens on the join expressions. This is usually not represented as a separate step in a query plan, in my experience at least, and is handled within the join operator itself. I agree that this would be equivalent in compute to having a separate projection in the plan.
 
I would be fine if we just want to support join on columns for this PR and look at supporting any expression as a separate issue. We can always work around the limitation by doing explicit projections before the join.

Reply:
For joins, the `column`=`column` case, the so called `equijoin` is so common and critical, I think it is important to make *super fast* and thus special cased. The more general expression case, while supported, is much less important than fast equi-joins, I think

I have seen both approaches (the join handles generic expressions, or the planner arranges to have only equijoins). For OUTER joins there is no choice but to do the evaluation in the join operator itself